frappe/email/oauth.py

```diff
@@ -0,0 +1,163 @@
+import base64
+from imaplib import IMAP4
+from poplib import POP3
+from smtplib import SMTP
+from typing import Union
+from urllib.parse import quote
+
+import frappe
+from frappe.integrations.google_oauth import GoogleOAuth
+from frappe.utils import get_request_site_address
+
+
+class OAuthenticationError(Exception):
+	pass
+
+
+class Oauth:
+	def __init__(
+		self,
+		conn: Union[IMAP4, POP3, SMTP],
+		email_account: str,
+		email: str,
+		access_token: str,
+		refresh_token: str,
+		service: str,
+		mechanism: str = "XOAUTH2",
+	) -> None:
+
+		self.email_account = email_account
+		self.email = email
+		self.service = service
+		self._mechanism = mechanism
+		self._conn = conn
+		self._access_token = access_token
+		self._refresh_token = refresh_token
+
+		self._validate()
+
+	def _validate(self) -> None:
+		if self.service != "GMail":
+			raise NotImplementedError(
+				f"Service {self.service} currently doesn't have oauth implementation."
+			)
+
+		if not self._refresh_token:
+			frappe.throw(
+				frappe._("Please Authorize OAuth."),
+				OAuthenticationError,
+				frappe._("OAuth Error"),
+			)
+
+	@property
+	def _auth_string(self) -> str:
+		return "user=%s\1auth=Bearer %s\1\1" % (self.email, self._access_token)
+
+	def connect(self, _retry: int = 0) -> None:
+		"""Connection method with retry on exception for Oauth"""
+		try:
+			if isinstance(self._conn, POP3):
+				res = self._connect_pop()
+
+				if not res.startswith(b"+OK"):
+					raise
```

Comment:
raising bare exception? Maybe we should log more info here.

Reply:
yep that's intentional ..it will get caught in the try except block for retrying/regenerating access token and if it again fails then we raise an OAuthenticationError